### PR TITLE
Special bug

### PR DESCRIPTION
**Pull request by lsgunth**
A couple more fixes starting with a bug in the special disk handling for the journal.

### CODE REVIEW

**Review by iomartin — APPROVED** [review]
Looks good to me

**Review by dlsloan — DISMISSED** [review]
Looks good to me!

**Comment by lsgunth** [comment]
Added one more fixup

File: md.py [hunk 1]
```diff
@@ -197,6 +197,8 @@ def __init__(self, md="md0", level=5, devs=None, ndisks=None,
                 raise MDInvalidArgumentError("Must not specify both disks and loop_disks")
             if size is None:
                 raise MDInvalidArgumentError("Must specify size with loop_disks")
+            #Placeholder that will be recreated in setup()
+            self.devs = [None] * ndisks
```
Comment (dlsloan) [inline md.py]: test3 --verify actually needs the disks to be created for zeroing

**Comment by lsgunth** [comment]
Ok revamped the disk zeroing, it should be much improved.

File: md.py [hunk 2]
```diff
@@ -134,6 +134,13 @@ class MDInvalidArgumentError(Exception):
     pass
 
 class MDInstance:
+    # Raid superblock v1.2 consumes (256 + 2 * ndisks) bytes at a 4k offset
```
Comment (dlsloan) [inline md.py]: Comment needs an update with the size change

File: md.py [hunk 3]
```diff
@@ -281,11 +293,30 @@ def get_num_disks(self):
...
+                               "bs=1M", f"count={self._size_to_zero}",
+                               "oflag=direct"],
+                              stderr=subprocess.DEVNULL)
+
+    def zero_all_disks(self, size_to_zero):
+        self._stop_and_create_disks()
+
+        count = size_to_zero + self.MAX_SUPERBLOCK_SZ
+        count = (count + (1 << 20) - 1) >> 20
+        self._size_to_zero = count
```
Comment (dlsloan) [inline md.py:306]: Should we add a dev size limit here in case we have very small ram disks or large zeroing ranges? Or id dd ok with count*bs > dev size?

Comment (lsgunth) [reply md.py:306]: It'll fail, but it's probably good to fail. If the device isn't at least that large the test will fail in other ways.

File: md.py (hunk 2 above)
Comment (lsgunth) [reply md.py]: Fixed

File: md.py (hunk 3 above)
Comment (dlsloan) [reply md.py:306]: We only need to zero size_to_zero / ndata + data_offset, so for small disks/large regions we're limiting the test size to a fraction of the disk, Might be a good idea to be able to test the end of the array as well as the beginning.

Comment (dlsloan) [reply md.py:306]: On second thought, we're using a spare disk as the veification scratch pad so end of disk tests will need special consideration anyway, this is good as is.

**Review by dlsloan — APPROVED** [review]

File: test_all [hunk 4]
```diff
@@ -27,6 +27,14 @@ def run_all_tests(env):
     run_test(["fio", "test2.fio"], env)
     run_test(["fio", "test_4jobs.fio"], env)
 
+def run_all_quick(env):
+    run_test(["./test1"], env)
+    run_test(["./test3", "--verify", "--quick"], env)
+    run_test(["./test3", "-d", "--quick"], env)
+    if not env or env.get("NO_RESHAPE") != "y":
+        run_test(["./test3", "-g", "--quick"], env)
+    run_test(["./test2"], env)
```
Comment (dlsloan) [inline test_all:36]: Big fan of this! I've basically been doing this manually for my testing

**Pull request merged**